http/http_response.hpp

```diff
@@ -116,37 +130,53 @@ struct Response
     {
         if (completed)
         {
-            BMCWEB_LOG_ERROR << "Response was ended twice";
+            BMCWEB_LOG_ERROR << this << " Response was ended twice";
             return;
         }
         completed = true;
-        BMCWEB_LOG_DEBUG << "calling completion handler";
+        BMCWEB_LOG_DEBUG << this << " calling completion handler";
         if (completeRequestHandler)
         {
-            BMCWEB_LOG_DEBUG << "completion handler was valid";
-            completeRequestHandler();
+            BMCWEB_LOG_DEBUG << this << " completion handler was valid";
+            completeRequestHandler(*this);
         }
     }
 
-    void end(std::string_view bodyPart)
+    bool isAlive()
+    {
+        return isAliveHelper && isAliveHelper();
+    }
+
+    void setCompleteRequestHandler(std::function<void(Response&)>&& handler)
     {
-        write(bodyPart);
-        end();
+        BMCWEB_LOG_DEBUG << this << " setting completion handler";
+        completeRequestHandler = std::move(handler);
     }
 
-    bool isAlive()
+    std::function<void(Response&)> releaseCompleteRequestHandler()
     {
-        return isAliveHelper && isAliveHelper();
+        BMCWEB_LOG_DEBUG << this << " releasing completion handler"
+                         << static_cast<bool>(completeRequestHandler);
+        std::function<void(Response&)> ret = completeRequestHandler;
+        completeRequestHandler = nullptr;
+        return ret;
     }
 
-    void setCompleteRequestHandler(std::function<void()> newHandler)
+    void setIsAliveHelper(std::function<bool()>&& handler)
     {
-        completeRequestHandler = std::move(newHandler);
+        isAliveHelper = std::move(handler);
+    }
+
+    std::function<bool()> releaseIsAliveHelper()
+    {
+        std::function<bool()> ret = std::move(isAliveHelper);
+        isAliveHelper = nullptr;
+        return ret;
     }
 
   private:
-    bool completed{};
-    std::function<void()> completeRequestHandler;
+    bool completed = false;
+    std::function<void(Response&)> completeRequestHandler;
```

Comment:
Are we sure this is the way to go? 
Have we added all the nullptrs where we set 'completeRequestHandler' that don't need to handle an extra response?

Is this clear enough question?

Reply:
> Are we sure this is the way to go? 

The codes are from Ed, I just help him split the patchset. If you have questions about the code, you can communicate with him.
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/47474

> Have we added all the nullptrs where we set 'completeRequestHandler' that don't need to handle an extra response?

I think we have added.